In [1]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [2]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [7]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

^C


### Test input output files

In [5]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [6]:
%%writefile output1.txt
12

Writing output1.txt


In [8]:
%%writefile input2.txt
0 9

Writing input2.txt


In [9]:
%%writefile output2.txt
9

Writing output2.txt


In [10]:
!./APlusB < input1.txt

12

In [11]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [15]:
!grep -xvFf output2.txt output1.txt

12


In [14]:
!grep -xvFf output1.txt output2.txt

9


In [12]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [13]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [16]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [18]:
%run sum.py

4 8
12


In [19]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [20]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [21]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [23]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [24]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [25]:
%%writefile output0.txt
0

Writing output0.txt


In [26]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [27]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [28]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [29]:
%%writefile output1.txt
90

Overwriting output1.txt


In [30]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [31]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [32]:
%%writefile output2.txt
90

Overwriting output2.txt


In [33]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [34]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [35]:
200000*200000

40000000000

In [36]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [37]:
!./max_pairwise_product_full < input3.txt

1345294336


In [38]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [39]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [40]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [41]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [42]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [43]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [44]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [45]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [53]:
%%writefile stress_test_generator.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cstdlib> // Para rand() y srand()
#include <ctime>   // Para seeding srand() con la hora actual

using namespace std;

// Función para generar un número entero aleatorio en un rango dado
int random_int(int min, int max) {
    return min + rand() % (max - min + 1);
}

// Función para calcular el producto máximo de dos elementos distintos de un vector
long long MaxPairwiseProduct(const vector<int>& numbers) {
    long long max_product = 0; // Inicializamos el máximo producto
    int n = numbers.size();    // Obtenemos el tamaño del vector

    // Comparar cada par de elementos del vector
    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
            if (first != second) { // Evitar multiplicar un número consigo mismo
                long long product = static_cast<long long>(numbers[first]) * numbers[second];
                if (product > max_product) {
                    max_product = product; // Actualizar si encontramos un producto mayor
                }
            }
        }
    }

    return max_product;
}

// Función para generar archivos de pruebas de estrés
void generate_stress_test(int start_index, int num_tests, int min_size, int max_size, int min_value, int max_value) {
    srand(time(nullptr)); // Inicializamos la semilla de números aleatorios

    // Generar 'num_tests' archivos de entrada y salida
    for (int i = start_index; i < start_index + num_tests; ++i) {
        int n = random_int(min_size, max_size); // Tamaño aleatorio para el vector

        // Crear el archivo de entrada
        ofstream input_file("input" + to_string(i) + ".txt");
        input_file << n << "\n"; // Escribimos el tamaño del vector

        vector<int> a(n);
        for (int j = 0; j < n; ++j) {
            a[j] = random_int(min_value, max_value); // Generamos valores aleatorios para el vector
            input_file << a[j] << " ";              // Escribimos cada valor en el archivo
        }
        input_file << "\n"; // Nueva línea al final del archivo
        input_file.close(); // Cerramos el archivo de entrada

        // Crear el archivo de salida con el resultado
        ofstream output_file("output" + to_string(i) + ".txt");
        output_file << MaxPairwiseProduct(a) << "\n"; // Escribimos el producto máximo en el archivo
        output_file.close(); // Cerramos el archivo de salida
    }
}

int main() {
    int start_index, num_tests;

    // Pedimos al usuario el índice inicial y la cantidad de pruebas a generar
    cout << "Ingrese el índice inicial para los archivos de entrada: ";
    cin >> start_index;
    cout << "Ingrese la cantidad de pruebas de estrés a crear: ";
    cin >> num_tests;

    // Llamamos a la función para generar las pruebas de estrés
    generate_stress_test(start_index, num_tests, 2, 1000, 0, 10000);

    return 0;
}



Overwriting stress_test_generator.cpp


In [54]:
# Compila el código de C++ con g++
!g++ stress_test_generator.cpp -o stress_test_generator

# Ejecuta el programa generado
!./stress_test_generator

# Ejemplo de entrada:
# Start index: 4 for file4
# numbers of tests: 8 (from 4 to 10)

Ingrese el índice inicial para los archivos de entrada: 4
Ingrese la cantidad de pruebas de estrés a crear: 8


In [55]:
##input4.txt with expected output4.txt
!cat input4.txt
!cat output4.txt

549
7481 7653 1874 2713 6531 1524 3960 9971 5119 8087 8584 6262 3020 6349 6106 5357 5343 5143 7337 6274 9331 529 7786 2084 8577 5849 6499 8734 9344 9517 316 6824 8249 3269 615 4779 4793 5654 4749 991 4819 4411 7253 7839 1838 4437 3195 8260 9581 531 4534 9990 2139 2319 2074 716 9248 9652 9450 8591 247 844 6493 8496 4113 8188 4353 9986 3841 9102 976 8661 4591 9308 7578 6429 3745 1852 5768 4404 3462 301 4393 5602 3699 7546 6318 2946 8277 6846 2615 8524 8769 187 7019 2881 8375 2450 3945 2215 1551 4921 1954 7221 4229 9533 4728 9053 1384 495 4535 4846 796 6 447 4496 7553 7844 8521 6908 5768 1136 5431 4536 1323 3528 8497 9698 5978 2441 2991 8608 7363 4946 5828 2670 4478 555 1722 6941 1051 6257 1786 2926 6263 3313 7422 4894 1156 7022 1801 6925 8158 8311 1460 559 2917 1035 1335 9974 3477 4326 8581 1918 9272 5487 4588 4828 7122 7389 1768 9252 4724 4634 2177 2065 7947 678 8039 181 7700 9840 7106 6936 9230 9646 7495 2146 1759 8830 2120 5236 4234 1779 8233 3506 7267 2820 9413 5467 1287 2260 4718 70

In [56]:
##input7.txt with expected output7.txt
!cat input7.txt
!cat output7.txt

559
9535 1310 2526 1340 77 630 4126 1561 3866 4383 4194 5658 3728 3233 8872 5868 9246 9557 2485 993 3300 6757 1693 5211 2171 5720 3713 4594 3048 3353 7194 3661 4663 798 5002 5819 1428 206 7380 6373 5668 1574 2031 9396 5886 1981 6342 5131 1538 9906 7203 4838 7741 8896 1127 9913 5694 4840 5585 8743 8193 2778 3482 3935 3577 9563 9754 6084 9769 8213 2457 6515 865 5567 6990 6751 7548 3331 2960 9086 3237 162 5002 977 136 6130 1968 5830 969 7553 5651 241 1410 9134 4176 6066 9775 5008 2149 9544 3220 5685 7137 5164 1251 4126 2993 9879 7458 5953 42 1773 7194 5045 3829 7330 2253 5798 4239 3222 3350 968 4542 4760 1180 8718 1904 955 4805 4054 1577 9104 817 8714 4268 3148 2840 7261 3026 1376 3214 3068 3149 1486 9192 6978 9896 1444 3854 4134 5746 7205 6181 287 3043 7362 84 4948 9396 4889 80 972 3992 1976 764 8260 5124 3604 6600 9229 6059 892 3376 286 2378 2567 8344 2273 5091 2197 7486 1915 480 3667 3281 4603 2107 3365 9551 1502 9333 709 3553 4404 2685 4317 2663 8889 9001 341 9196 6138 1233 2571 6425 

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [57]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [58]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [59]:
199999 * 200000

39999800000

In [60]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [61]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [62]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [63]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [64]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	4m58.257s
user	4m56.107s
sys	0m0.075s


In [65]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m58.080s
user	4m55.767s
sys	0m0.083s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [66]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [67]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [68]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [69]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [70]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [71]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

572


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177092


In [ ]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [72]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [73]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [74]:
10999 * 11000

120989000

In [75]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.845s
user	0m0.840s
sys	0m0.003s


## Python

In [76]:
%%writefile max_pairwise_product.py
def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]

                if aux_product > max_product:
                  max_product = aux_product

    return max_product

n = int(input())

numbers = [int(i) for i in input().split()]

print(max_pairwise_product(numbers))


Writing max_pairwise_product.py


In [77]:
!time python3 max_pairwise_product.py < input11.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product.py", line 19, in <module>
    print(max_pairwise_product(numbers))
  File "/content/max_pairwise_product.py", line 10, in max_pairwise_product
    if aux_product > max_product:
KeyboardInterrupt
^C


the code runs for more than 30min.However, the code did not finish executing, this is because Python is much slower than C++ to carry out the instructions.

# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [78]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [79]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [129]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>  // Para std::cin y std::cout
#include <vector>    // Para std::vector
#include <cstdlib>   // Para rand()
#include <ctime>     // Para srand() y time()

// Función para calcular el producto máximo de un par de elementos
long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
    int n = numbers.size();

    // Recorre todos los pares posibles de elementos
    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            long long aux_product = static_cast<long long>(numbers[first]) * numbers[second];
            // Actualiza el producto máximo si se encuentra un producto mayor
            if (aux_product > max_product) {
                max_product = aux_product;
            }
        }
    }

    return max_product;
}

int main() {
    int n;

    // Solicita al usuario el tamaño del vector
    std::cout << "Enter the size of the vector (n): ";
    std::cin >> n;

    std::vector<int> numbers(n);  // Crea un vector de tamaño n

    // Solicita al usuario los elementos del vector
    std::cout << "Enter the elements of the vector:\n";
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];  // Lee cada elemento del vector
    }

    // Calcula y muestra el producto máximo de pares
    std::cout << "Max pairwise product: " << MaxPairwiseProduct(numbers) << "\n";

    return 0;
}



Overwriting max_pairwise_product_long.cpp


### Estimate and verify the maximun n that can be solved in 1 second



In [127]:
%%writefile max_n_in_a_second.cpp
#include <iostream>
#include <vector>
#include <chrono>

// Función para calcular el producto máximo de un par de elementos
long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
    int n = numbers.size();

    // Recorre todos los pares posibles de elementos
    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            long long aux_product = static_cast<long long>(numbers[first]) * numbers[second];
            // Actualiza el producto máximo si se encuentra un producto mayor
            if (aux_product > max_product) {
                max_product = aux_product;
            }
        }
    }

    return max_product;
}

int main() {
    int n = 10;  // Valor inicial de n
    auto start_time = std::chrono::high_resolution_clock::now();  // Marca el inicio del tiempo total

    // Bucle para encontrar el valor máximo de n que puede ser resuelto en 1 segundo
    while (true) {
        // Genera los datos de entrada de tamaño n (todos los valores son 1)
        std::vector<int> input_data(n, 1);

        // Mide el tiempo de ejecución de la función MaxPairwiseProduct
        auto start = std::chrono::high_resolution_clock::now();
        MaxPairwiseProduct(input_data);  // Llama a la función para calcular el producto máximo
        auto end = std::chrono::high_resolution_clock::now();

        // Calcula el tiempo transcurrido en milisegundos
        double elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;

        // Si el tiempo transcurrido es mayor o igual a 1 segundo, detiene el bucle
        if (elapsed_time >= 1.0) {
            break;
        }

        // Doble el valor de n para el siguiente ciclo
        n *= 2;
    }

    auto end_time = std::chrono::high_resolution_clock::now();  // Marca el final del tiempo total
    double total_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count() / 1000.0;

    // Muestra el valor máximo de n que puede resolverse en 1 segundo y el tiempo total de ejecución
    std::cout << "Maximum n that can be solved in 1 second: " << n << std::endl;
    std::cout << "Total execution time: " << total_time << " seconds" << std::endl;

    return 0;
}


Overwriting max_n_in_a_second.cpp


In [128]:
!g++ -o max_n_in_a_second max_n_in_a_second.cpp
!./max_n_in_a_second

Maximum n that can be solved in 1 second: 20480
Total execution time: 1.811 seconds


In [120]:
%%writefile  stress_test_of_size_n.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>  // Para la semilla de srand()

using namespace std;

// Función para generar un número entero aleatorio en un rango especificado
int random_int(int min, int max) {
    return min + rand() % (max - min + 1);
}

// Función para generar una prueba de estrés de tamaño n
void generate_stress_test(int n, int min_value, int max_value) {
    srand(time(nullptr)); // Inicializa la semilla de srand con la hora actual

    // Crea el archivo de entrada
    ofstream input_file("stress_test_of_size_n.txt");

    // Escribe el tamaño del vector en el archivo
    input_file << n << "\n";

    // Genera valores aleatorios y los escribe en el archivo
    for (int i = 0; i < n; ++i) {
        int random_value = random_int(min_value, max_value);
        input_file << random_value << " ";  // Escribe el elemento del vector
    }

    input_file << "\n";  // Marca el final del vector
    input_file.close();  // Cierra el archivo
}

int main() {
    int n;

    // Solicita al usuario el tamaño del vector
    cout << "Enter the size of the vector (n): ";
    cin >> n;

    int min_value = 0;     // Valor mínimo para los enteros aleatorios
    int max_value = 200000; // Valor máximo para los enteros aleatorios

    // Genera una prueba de estrés con el tamaño especificado
    generate_stress_test(n, min_value, max_value);

    return 0;
}




Overwriting stress_test_of_size_n.cpp


In [121]:
#Create stress test of size n = 20480
!g++ -o stress_test_of_size_n stress_test_of_size_n.cpp
!./stress_test_of_size_n

Enter the size of the vector (n): 20480


In [130]:
#Verify that max_wisepair_product_long can run an input of size n = 20480 in a second
!g++ -o max_pairwise_product_long max_pairwise_product_long.cpp
!time ./max_pairwise_product_long < stress_test_of_size_n.txt

Enter the size of the vector (n): Enter the elements of the vector:
Max pairwise product: 39992400361

real	0m2.141s
user	0m2.064s
sys	0m0.004s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [132]:
%%writefile max_pairwise_product_long_sorting.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>  // Para std::log

// Implementación de introsort
template <typename RandomIt>
void introsort(RandomIt first, RandomIt last, int depth_limit) {
    while (last - first > 16) {  // Cuando el rango es mayor a 16 elementos
        if (depth_limit == 0) {
            // Si el límite de profundidad se ha alcanzado, usa un heap sort
            std::make_heap(first, last);
            std::sort_heap(first, last);
            return;
        }
        --depth_limit;

        // Particiona el rango y elige un pivote
        RandomIt pivot = std::partition(first, last, [=](const auto& em) { return em < *(last - 1); });
        std::swap(*pivot, *(last - 1));  // Intercambia el pivote con el último elemento
        introsort(first, pivot, depth_limit);  // Llama recursivamente sobre la primera mitad
        first = pivot + 1;  // Continuamos sobre la segunda mitad
    }
    // Si el rango es pequeño, utiliza un simple sort
    std::sort(first, last);
}

// Función envoltorio para usar introsort
template <typename RandomIt>
void introsort(RandomIt first, RandomIt last) {
    // Se calcula el límite de profundidad en función de la longitud del rango
    introsort(first, last, 2 * std::log(last - first));
}

// Función para calcular el producto máximo de pares
long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    // Creamos una copia del vector y la ordenamos
    std::vector<int> sorted_numbers = numbers;
    introsort(sorted_numbers.begin(), sorted_numbers.end());

    int n = sorted_numbers.size();
    // El producto máximo es el de los dos números más grandes
    return static_cast<long long>(sorted_numbers[n - 1]) * sorted_numbers[n - 2];
}

int main() {
    int n;

    // Solicita al usuario el tamaño del vector
    std::cin >> n;

    std::vector<int> numbers(n);

    // Lee los elementos del vector
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    // Calcula y muestra el producto máximo de pares
    std::cout << MaxPairwiseProduct(numbers) << "\n";

    return 0;
}


Overwriting max_pairwise_product_long_sorting.cpp


Test with inputs 1 and 2



In [133]:
!g++ -o max_pairwise_product_long_sorting max_pairwise_product_long_sorting.cpp
!./max_pairwise_product_long_sorting < input1.txt
!./max_pairwise_product_long_sorting < input2.txt

90
90


Test and time it with input 3

In [134]:
!g++ -o max_pairwise_product_long_sorting max_pairwise_product_long_sorting.cpp
!time ./max_pairwise_product_long_sorting < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


20 stress tests

In [136]:
#stress tests randomly created with index from 13 to 32 for files
!g++ stress_test_generator.cpp -o stress_test_generator
!./stress_test_generator

#Start index = 13
# Number of tests: 20

Ingrese el índice inicial para los archivos de entrada: 13
Ingrese la cantidad de pruebas de estrés a crear: 20


In [139]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full

# Bucle para procesar los archivos desde input13.txt hasta input32.txt
for i in {13..32}
do
    # Ejecutar el programa con el archivo de entrada correcto
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    echo "Testing on input$i"

    # Medir el tiempo de ejecución
    time ./max_pairwise_product_full < input$i.txt

    # Comparar la salida generada con la salida esperada
    echo ""

    # Verificar la diferencia entre el archivo de salida esperado y el generado
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [140]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Testing on input13
99950006

real	0m0.009s
user	0m0.007s
sys	0m0.002s

Testing on input14
94438899

real	0m0.002s
user	0m0.000s
sys	0m0.002s

Testing on input15
98624680

real	0m0.002s
user	0m0.002s
sys	0m0.000s

Testing on input16
99161643

real	0m0.004s
user	0m0.003s
sys	0m0.001s

Testing on input17
99750024

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input18
73533650

real	0m0.003s
user	0m0.001s
sys	0m0.001s

Testing on input19
99430572

real	0m0.006s
user	0m0.004s
sys	0m0.001s

Testing on input20
99750066

real	0m0.005s
user	0m0.004s
sys	0m0.001s

Testing on input21
98575100

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input22
99810088

real	0m0.009s
user	0m0.009s
sys	0m0.000s

Testing on input23
98218008

real	0m0.004s
user	0m0.003s
sys	0m0.000s

Testing on input24
99800099

real	0m0.008s
user	0m0.006s
sys	0m0.002s

Testing on input25
99440735

real	0m0.007s
user	0m0.006s
sys	0m0.001s

Testing on input26
87694731

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testin

Count executed instructions in code and verify with inputs 1,2 and 3


In [142]:
%%writefile max_pairwise_product_long_sorting_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath> // Para std::log

// Variable global para contar las instrucciones ejecutadas
long long execints = 0LL;

// Implementación del introsort
template <typename RandomIt>
void introsort(RandomIt first, RandomIt last, int depth_limit) {
    while (last - first > 16) {
        if (depth_limit == 0) {
            std::make_heap(first, last);
            std::sort_heap(first, last);
            return;
        }
        --depth_limit;
        RandomIt pivot = std::partition(first, last, [=](const auto& em) { return em < *(last - 1); });
        std::swap(*pivot, *(last - 1));
        introsort(first, pivot, depth_limit);
        first = pivot + 1;
    }
    std::sort(first, last);
}

// Función envolvente para usar introsort
template <typename RandomIt>
void introsort(RandomIt first, RandomIt last) {
    introsort(first, last, 2 * std::log(last - first));
}

// Función para calcular el producto máximo de pares
long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    std::vector<int> sorted_numbers = numbers;
    introsort(sorted_numbers.begin(), sorted_numbers.end()); // Ordena el vector

    int n = sorted_numbers.size();
    execints += 2; // Para la asignación y la comprobación del tamaño
    execints += 2 * std::log(n); // Para el proceso de ordenación

    // Retorna el producto de los dos elementos más grandes
    return (long long)sorted_numbers[n - 1] * sorted_numbers[n - 2];
}

int main() {
    int n;
    std::cin >> n;
    execints += 1; // Para leer n

    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
        execints += 2; // Para leer el número y el incremento de i
    }

    // Imprime el número total de instrucciones ejecutadas
    std::cout << execints << "\n";

    return 0;
}


Overwriting max_pairwise_product_long_sorting_execints.cpp


In [143]:
#Compute the time using some of the input files
!g++ max_pairwise_product_long_sorting_execints.cpp -o max_pairwise_product_long_sorting_execints
!time ./max_pairwise_product_long_sorting_execints < input11.txt  > output11.txt
!time ./max_pairwise_product_long_sorting_execints < input25.txt  > output25.txt
!time ./max_pairwise_product_long_sorting_execints < input32.txt  > output32.txt


real	0m0.069s
user	0m0.066s
sys	0m0.003s

real	0m0.004s
user	0m0.001s
sys	0m0.003s

real	0m0.009s
user	0m0.003s
sys	0m0.001s


Calculate estimated time by instruction and number of instructions by second

In [145]:
# Read the content of output files
def read_output(file_name):
    with open(file_name, 'r') as file:
        return file.read().strip()

# Read output content from respective files
output1_content = read_output('output11.txt')
output2_content = read_output('output25.txt')
output3_content = read_output('output32.txt')

# Parse the number of executed instructions
executed_instructions_1 = int(output1_content)
executed_instructions_2 = int(output2_content)
executed_instructions_3 = int(output3_content)

# Real-time measurements in seconds
real_time_1 = 0.073
real_time_2 = 0.003
real_time_3 = 0.004

# Compute Estimated Time per Instruction (ETI) and Instructions per Second (IPS)
def calculate_metrics(real_time, executed_instructions):
    eti = real_time / executed_instructions
    ips = executed_instructions / real_time
    return eti, ips

eti_1, ips_1 = calculate_metrics(real_time_1, executed_instructions_1)
eti_2, ips_2 = calculate_metrics(real_time_2, executed_instructions_2)
eti_3, ips_3 = calculate_metrics(real_time_3, executed_instructions_3)

# Print the results
def print_metrics(output_number, eti, ips):
    print(f"Output {output_number}:")
    print(f"ETI: {eti:.9f} s/instruction")
    print(f"IPS: {ips:.0f} instructions/second")

print_metrics(1, eti_1, ips_1)
print("\n")
print_metrics(2, eti_2, ips_2)
print("\n")
print_metrics(3, eti_3, ips_3)



Output 1:
ETI: 0.000000182 s/instruction
IPS: 5479466 instructions/second


Output 2:
ETI: 0.000001811 s/instruction
IPS: 552333 instructions/second


Output 3:
ETI: 0.000002586 s/instruction
IPS: 386750 instructions/second


### Estimate and verify the maximun n that can be solved in 1 second

In [146]:
%%writefile max_n_in_a_second_sorting.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>    // Include cmath for std::log
#include <chrono>   // Include chrono for timing

// Implement introsort
template <typename RandomIt>
void introsort(RandomIt first, RandomIt last, int depth_limit) {
    while (last - first > 16) {
        if (depth_limit == 0) {
            std::make_heap(first, last);
            std::sort_heap(first, last);
            return;
        }
        --depth_limit;
        RandomIt pivot = std::partition(first, last, [=](const auto& em) { return em < *(last - 1); });
        std::swap(*pivot, *(last - 1));
        introsort(first, pivot, depth_limit);
        first = pivot + 1;
    }
    std::sort(first, last);
}

// Wrapper function to use introsort
template <typename RandomIt>
void introsort(RandomIt first, RandomIt last) {
    introsort(first, last, 2 * std::log(last - first));
}

// Function to calculate the maximum pairwise product
long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    std::vector<int> sorted_numbers = numbers;
    introsort(sorted_numbers.begin(), sorted_numbers.end());

    int n = sorted_numbers.size();
    return (long long)sorted_numbers[n - 1] * sorted_numbers[n - 2];
}

// Function to measure execution time for a given input size
double measureTimeForN(int n) {
    std::vector<int> numbers(n, 1); // Generate input with all elements as 1
    auto start = std::chrono::high_resolution_clock::now(); // Start the timer
    MaxPairwiseProduct(numbers); // Call the function
    auto end = std::chrono::high_resolution_clock::now(); // Stop the timer
    std::chrono::duration<double> elapsed_seconds = end - start; // Calculate duration
    return elapsed_seconds.count(); // Return the duration in seconds
}

// Function to find the maximum n that can be solved within a given time limit
int findMaxNWithinTimeLimit(double time_limit) {
    int low = 100; // Initial low value of n
    int high = low; // Initial high value of n

    // Increase high until execution time exceeds the time limit
    while (measureTimeForN(high) <= time_limit) {
        low = high;
        high *= 2;
    }

    // Binary search to find the maximum n within the time limit
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (measureTimeForN(mid) <= time_limit) {
            low = mid + 1; // Move to the right half
        } else {
            high = mid - 1; // Move to the left half
        }
    }

    return high; // Return the last successful value of n
}

int main() {
    double time_limit = 1.0; // Time limit in seconds
    int max_n = findMaxNWithinTimeLimit(time_limit);
    std::cout << "Maximum n that can be solved in 1 second: " << max_n << std::endl;
    return 0;
}


Writing max_n_in_a_second_sorting.cpp


In [147]:
!g++ -o max_n_in_a_second_sorting max_n_in_a_second_sorting.cpp
!./max_n_in_a_second_sorting

Maximum n that can be solved in 1 second: 583646


In [149]:
#Create stress test of size n = 550000
!g++ -o stress_test_of_size_n stress_test_of_size_n.cpp
!./stress_test_of_size_n

Enter the size of the vector (n): 550000


In [150]:
#Verify the max_pairwise_product_long_sorting can run stress test of size n = 550000

!g++ -omax_pairwise_product_long_sorting max_pairwise_product_long_sorting.cpp
!time ./max_pairwise_product_long_sorting < stress_test_of_size_n.txt

40000000000

real	0m0.779s
user	0m0.742s
sys	0m0.006s


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [151]:
%%writefile max_pairwise_product_long_efficient_execints.cpp
#include <iostream>
#include <vector>
#include <utility> // Include for std::pair

std::pair<long long, long long> MaxPairwiseProduct(const std::vector<int>& numbers, long long& execints) {
    long long max_product = 0; // Initialize max product
    execints += 1; // Increment instruction count for initialization

    long long second_max = 0; // Initialize second max
    execints += 1; // Increment instruction count for initialization

    for (int i = 0; i < numbers.size(); ++i) {
        execints += 3; // Increment instruction count for loop operation
        execints += 2; // Increment instruction count for comparison

        // Update max product and second max values
        if (numbers[i] > max_product) {
            second_max = max_product; // Update second max
            execints += 1; // Increment instruction count
            max_product = numbers[i]; // Update max product
            execints += 1; // Increment instruction count
        } else if (numbers[i] > second_max) {
            second_max = numbers[i]; // Update second max
            execints += 1; // Increment instruction count
        }
    }

    return std::make_pair(max_product, second_max); // Return the pair (max product, second max)
}

int main() {
    long long execints = 0; // Initialize instruction count
    int n; // Declare the number of elements
    execints += 1; // Increment instruction count for variable declaration
    std::cin >> n; // Input the number of elements
    execints += 1; // Increment instruction count for input

    std::vector<int> numbers(n); // Initialize the vector with size n
    execints += n + 2; // Increment instruction count for vector initialization

    for (int i = 0; i < n; ++i) {
        execints += 3; // Increment instruction count for loop operation
        execints += 2; // Increment instruction count for comparison
        std::cin >> numbers[i]; // Input each number into the vector
        execints += 1; // Increment instruction count for input
    }

    auto result = MaxPairwiseProduct(numbers, execints); // Get the pair (max product, second max)
    execints += 1; // Increment instruction count for calling MaxPairwiseProduct
    std::cout << "Number of executed instructions: " << execints << std::endl; // Output the instruction count
    return 0; // Return 0 to indicate successful completion
}


Writing max_pairwise_product_long_efficient_execints.cpp


Test with inputs 1 and 2



In [152]:
!g++ -o max_pairwise_product_long_efficient_execints max_pairwise_product_long_efficient_execints.cpp
!time ./max_pairwise_product_long_efficient_execints < input1.txt
!time ./max_pairwise_product_long_efficient_execints < input2.txt

Number of executed instructions: 147

real	0m0.003s
user	0m0.000s
sys	0m0.003s
Number of executed instructions: 130

real	0m0.003s
user	0m0.001s
sys	0m0.001s


Test and time it with input 3

In [153]:
!g++ -o max_pairwise_product_long_efficient_execints max_pairwise_product_long_efficient_execints.cpp
!time ./max_pairwise_product_long_efficient_execints < input3.txt

Number of executed instructions: 87

real	0m0.003s
user	0m0.001s
sys	0m0.002s


20 stress tests

In [154]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full

# Bucle para procesar los archivos desde input13.txt hasta input32.txt
for i in {13..32}
do
    # Ejecutar el programa con el archivo de entrada correcto
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    echo "Testing on input$i"

    # Medir el tiempo de ejecución
    time ./max_pairwise_product_full < input$i.txt

    # Comparar la salida generada con la salida esperada
    echo ""

    # Verificar la diferencia entre el archivo de salida esperado y el generado
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [155]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Testing on input13
99950006

real	0m0.009s
user	0m0.009s
sys	0m0.001s

Testing on input14
94438899

real	0m0.002s
user	0m0.000s
sys	0m0.002s

Testing on input15
98624680

real	0m0.003s
user	0m0.000s
sys	0m0.003s

Testing on input16
99161643

real	0m0.004s
user	0m0.002s
sys	0m0.002s

Testing on input17
99750024

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input18
73533650

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input19
99430572

real	0m0.005s
user	0m0.005s
sys	0m0.000s

Testing on input20
99750066

real	0m0.005s
user	0m0.004s
sys	0m0.001s

Testing on input21
98575100

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input22
99810088

real	0m0.010s
user	0m0.008s
sys	0m0.001s

Testing on input23
98218008

real	0m0.004s
user	0m0.003s
sys	0m0.001s

Testing on input24
99800099

real	0m0.008s
user	0m0.009s
sys	0m0.000s

Testing on input25
99440735

real	0m0.007s
user	0m0.006s
sys	0m0.001s

99440735
Testing on input26
87694731

real	0m0.002s
user	0m0.001s
sys	0m0.001

Count executed instructions in code and verify with inputs 1,2 and 3


In [156]:
!g++ -o max_pairwise_product_long_efficient_execints max_pairwise_product_long_efficient_execints.cpp
!./max_pairwise_product_long_efficient_execints < input1.txt
!./max_pairwise_product_long_efficient_execints < input2.txt
!./max_pairwise_product_long_efficient_execints < input3.txt

Number of executed instructions: 147
Number of executed instructions: 130
Number of executed instructions: 87


Calculate estimated time by instruction and number of instructions by second

In [157]:
!g++ max_pairwise_product_long_sorting_execints.cpp -o max_pairwise_product_long_sorting_execints
!time ./max_pairwise_product_long_sorting_execints < input12.txt  > output12.txt
!time ./max_pairwise_product_long_sorting_execints < input27.txt  > output27.txt
!time ./max_pairwise_product_long_sorting_execints < input31.txt  > output31.txt


real	0m0.123s
user	0m0.115s
sys	0m0.002s

real	0m0.005s
user	0m0.001s
sys	0m0.002s

real	0m0.005s
user	0m0.002s
sys	0m0.002s


In [158]:
with open('output12.txt', 'r') as file:
    output1_content = file.read().strip()

with open('output27.txt', 'r') as file:
    output2_content = file.read().strip()

with open('output31.txt', 'r') as file:
    output3_content = file.read().strip()

# Parse the number of executed instructions
executed_instructions_1 = int(output1_content)
executed_instructions_2 = int(output2_content)
executed_instructions_3 = int(output3_content)

# Real-time measurements in seconds
real_time_1 = 0.066
real_time_2 = 0.004
real_time_3 = 0.005

# Compute Estimated Time per Instruction (ETI) and Instructions per Second (IPS)
eti_1 = real_time_1 / executed_instructions_1
ips_1 = executed_instructions_1 / real_time_1

eti_2 = real_time_2 / executed_instructions_2
ips_2 = executed_instructions_2 / real_time_2

eti_3 = real_time_3 / executed_instructions_3
ips_3 = executed_instructions_3 / real_time_3

# Print the results
print("Output 1:")
print("ETI: {:.9f} s/instruction".format(eti_1))
print("IPS: {:.0f} instructions/second".format(ips_1))

print("\nOutput 2:")
print("ETI: {:.9f} s/instruction".format(eti_2))
print("IPS: {:.0f} instructions/second".format(ips_2))

print("\nOutput 3:")
print("ETI: {:.9f} s/instruction".format(eti_3))
print("IPS: {:.0f} instructions/second".format(ips_3))

Output 1:
ETI: 0.000000165 s/instruction
IPS: 6060621 instructions/second

Output 2:
ETI: 0.000003249 s/instruction
IPS: 307750 instructions/second

Output 3:
ETI: 0.000004163 s/instruction
IPS: 240200 instructions/second


### Estimate and verify the maximun n that can be solved in 1 second

In [159]:
%%writefile max_n_in_a_second_long_efficient.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <random>
#include <chrono>

// Function to generate a random integer in a specified range
int random_int(int min, int max) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis(min, max);
    return dis(gen);
}

// Function to generate input data of size n
void generate_input_data(int n, int min_value, int max_value) {
    std::ofstream input_file("input.txt");
    input_file << n << "\n"; // Output vector size

    // Generate n random values and write them to the file
    for (int j = 0; j < n; ++j) {
        int num = random_int(min_value, max_value); // Generate random values for vector
        input_file << num << " "; // Write vector element
    }
    input_file << "\n"; // End of vector
    input_file.close();
}

// Function to calculate the max pairwise product
std::pair<long long, long long> MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0;
    long long second_max = 0;

    // Find the max product and second max in the vector
    for (int i = 0; i < numbers.size(); ++i) {
        if (numbers[i] > max_product) {
            second_max = max_product;
            max_product = numbers[i];
        } else if (numbers[i] > second_max) {
            second_max = numbers[i];
        }
    }

    return std::make_pair(max_product, second_max);
}

int main() {
    int n = 10;               // Initial size of the input vector
    int min_value = 1;        // Minimum value for random numbers
    int max_value = 1000000;  // Maximum value for random numbers
    int max_n = 1000000;      // Maximum value of n to try
    auto start_time = std::chrono::high_resolution_clock::now();

    // Loop to find the maximum n that can be solved in 1 second
    while (n <= max_n) {
        // Generate input data for current n
        generate_input_data(n, min_value, max_value);

        // Measure the time taken to read and process the data
        auto start = std::chrono::high_resolution_clock::now();

        // Read input data from file
        std::ifstream input_file("input.txt");
        int num;
        std::vector<int> input_data;
        input_file >> num; // Read the size of the input
        input_data.reserve(num); // Reserve memory for input data
        while (input_file >> num) {
            input_data.push_back(num); // Read each number into the vector
        }
        input_file.close();

        // Calculate the pairwise product
        MaxPairwiseProduct(input_data);

        // Measure the time taken for the computation
        auto end = std::chrono::high_resolution_clock::now();
        double elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;
        std::cout << "n = " << n << ", Time taken: " << elapsed_time << " seconds" << std::endl;

        // Stop if the computation time exceeds 1 second
        if (elapsed_time >= 1.0) {
            break;
        }
        n *= 2; // Double the size of the input for the next iteration
    }

    // Output the maximum n that can be solved in 1 second
    std::cout << "Maximum n that can be solved in 1 second: " << n << std::endl;

    return 0;
}


Writing max_n_in_a_second_long_efficient.cpp


In [160]:
!g++ -o max_n_in_a_second_long_efficient max_n_in_a_second_long_efficient.cpp
!./max_n_in_a_second_long_efficient

n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0 seconds
n = 10240, Time taken: 0.001 seconds
n = 20480, Time taken: 0.002 seconds
n = 40960, Time taken: 0.005 seconds
n = 81920, Time taken: 0.01 seconds
n = 163840, Time taken: 0.021 seconds
n = 327680, Time taken: 0.043 seconds
n = 655360, Time taken: 0.107 seconds
Maximum n that can be solved in 1 second: 1310720


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [162]:
%%writefile max_pairwise_product_super_efficient.cpp
#include <iostream>

int main() {
    int n;
    std::cin >> n;  // Lee el tamaño de la lista de números

    // Inicializa las variables para almacenar el mayor y segundo mayor número
    long long largest = 0, second_largest = 0;

    // Bucle para procesar los números
    for (int i = 0; i < n; ++i) {
        int number;
        std::cin >> number;  // Lee un número

        // Actualiza el mayor y el segundo mayor número
        if (number > largest) {
            second_largest = largest;  // El antiguo mayor se convierte en el segundo mayor
            largest = number;          // Actualiza el mayor
        } else if (number > second_largest) {
            second_largest = number;   // Actualiza el segundo mayor si es mayor que el actual
        }
    }

    // Calcula el producto máximo por pares usando los dos mayores números
    long long max_product = largest * second_largest;

    // Imprime el producto máximo
    std::cout << max_product << "\n";

    return 0;  // Finaliza el programa
}


Overwriting max_pairwise_product_super_efficient.cpp


Test with inputs 1 and 2




In [163]:
!g++ -o max_pairwise_product_super_efficient max_pairwise_product_super_efficient.cpp
!time ./max_pairwise_product_super_efficient < input1.txt
!time ./max_pairwise_product_super_efficient < input2.txt

90

real	0m0.003s
user	0m0.000s
sys	0m0.002s
90

real	0m0.003s
user	0m0.001s
sys	0m0.002s


Test and time it with input 3

In [164]:
!time ./max_pairwise_product_super_efficient < input3.txt

40000000000

real	0m0.021s
user	0m0.001s
sys	0m0.002s


20 stress tests

In [166]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full

# Bucle para procesar los archivos desde input13.txt hasta input32.txt
for i in {13..32}
do
    # Ejecutar el programa con el archivo de entrada correcto
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    echo "Testing on input$i"

    # Medir el tiempo de ejecución
    time ./max_pairwise_product_full < input$i.txt

    # Comparar la salida generada con la salida esperada
    echo ""

    # Verificar la diferencia entre el archivo de salida esperado y el generado
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [167]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Testing on input13
99950006

real	0m0.008s
user	0m0.006s
sys	0m0.002s

Testing on input14
94438899

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input15
98624680

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input16
99161643

real	0m0.004s
user	0m0.003s
sys	0m0.001s

Testing on input17
99750024

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testing on input18
73533650

real	0m0.002s
user	0m0.002s
sys	0m0.000s

Testing on input19
99430572

real	0m0.005s
user	0m0.003s
sys	0m0.002s

Testing on input20
99750066

real	0m0.006s
user	0m0.004s
sys	0m0.001s

Testing on input21
98575100

real	0m0.002s
user	0m0.002s
sys	0m0.000s

Testing on input22
99810088

real	0m0.009s
user	0m0.008s
sys	0m0.001s

Testing on input23
98218008

real	0m0.004s
user	0m0.002s
sys	0m0.001s

Testing on input24
99800099

real	0m0.008s
user	0m0.006s
sys	0m0.002s

Testing on input25
99440735

real	0m0.007s
user	0m0.006s
sys	0m0.001s

99440735
Testing on input26
87694731

real	0m0.002s
user	0m0.001s
sys	0m0.001

Count executed instructions in code and verify with inputs 1,2 and 3

In [169]:
%%writefile max_pairwise_product_super_efficient_execints.cpp
#include <iostream>

int main() {
    int n;
    std::cin >> n;  // Lee el número de elementos en la lista de números

    // Inicializa las variables para almacenar el mayor y el segundo mayor número
    long long largest = 0, second_largest = 0;

    // Contador para el número de instrucciones ejecutadas
    long long execints = 0;

    // Bucle para procesar los números
    for (int i = 0; i < n; ++i) {
        execints += 3; // Incremento por la inicialización de la variable y la comprobación de la condición del bucle

        int number;
        std::cin >> number;  // Lee un número de la entrada
        execints += 2; // Incremento por la operación de entrada (leer número)

        // Compara el número con los dos mayores y actualiza las variables correspondientes
        execints += 4; // Incremento por las comparaciones y asignaciones dentro del bucle
        if (number > largest) {
            second_largest = largest; // El mayor número pasa a ser el segundo mayor
            largest = number; // Actualiza el mayor número
            execints += 2; // Incremento por las asignaciones de `second_largest` y `largest`
        } else if (number > second_largest) {
            second_largest = number; // Actualiza el segundo mayor número
            execints += 1; // Incremento por la asignación de `second_largest`
        }
    }

    // Calcula el producto máximo por pares
    long long max_product = largest * second_largest;
    execints += 1; // Incremento por la multiplicación

    // Muestra el producto máximo por pares
    std::cout << "Maximum pairwise product: " << max_product << "\n";
    execints += 2; // Incremento por la operación de salida (mostrar el resultado y el salto de línea)

    // Muestra el número de instrucciones ejecutadas
    std::cout << "Number of executed instructions: " << execints << std::endl;
    execints += 2; // Incremento por la operación de salida (mostrar el número de instrucciones y el salto de línea)

    return 0; // Finaliza el programa
}


Writing max_pairwise_product_super_efficient_execints.cpp


In [170]:
!g++ -o max_pairwise_product_super_efficient_execints max_pairwise_product_super_efficient_execints.cpp
!./max_pairwise_product_super_efficient_execints < input1.txt
!./max_pairwise_product_super_efficient_execints < input2.txt
!./max_pairwise_product_super_efficient_execints < input3.txt

Maximum pairwise product: 90
Number of executed instructions: 113
Maximum pairwise product: 90
Number of executed instructions: 96
Maximum pairwise product: 40000000000
Number of executed instructions: 65


Calculate estimated time by instruction and number of instructions by second

In [171]:
!g++ max_pairwise_product_super_efficient_execints.cpp -o max_pairwise_product_super_efficient_execints
!time ./max_pairwise_product_super_efficient_execints < input12.txt  > output12.txt
!time ./max_pairwise_product_super_efficient_execints < input27.txt  > output27.txt
!time ./max_pairwise_product_super_efficient_execints < input31.txt  > output31.txt


real	0m0.062s
user	0m0.060s
sys	0m0.001s

real	0m0.005s
user	0m0.003s
sys	0m0.000s

real	0m0.003s
user	0m0.003s
sys	0m0.000s


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [172]:
%%writefile max_n_in_a_second_super_efficient.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <random>
#include <chrono>
#include <limits>

// Función para generar un número entero aleatorio dentro de un rango especificado
int random_int(int min, int max) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(min, max);
    return dis(gen);
}

// Función para generar los datos de entrada con un tamaño n
void generate_input_data(int n, int min_value, int max_value) {
    std::ofstream input_file("input.txt");
    input_file << n << "\n"; // Salida del tamaño del vector

    // Genera números aleatorios para el vector
    for (int j = 0; j < n; ++j) {
        int num = random_int(min_value, max_value); // Genera un valor aleatorio para el vector
        input_file << num << " "; // Escribe el elemento en el archivo
    }
    input_file << "\n"; // Fin del vector
    input_file.close();
}

// Función para calcular el producto máximo de pares de manera eficiente
long long super_efficient_max_pairwise_product(const std::vector<int>& numbers) {
    long long largest = std::numeric_limits<long long>::min(); // Inicializa con el valor más pequeño posible
    long long second_largest = std::numeric_limits<long long>::min(); // Inicializa con el valor más pequeño posible

    // Recorre los números y actualiza los dos números más grandes
    for (int num : numbers) {
        if (num > largest) {
            second_largest = largest;
            largest = num;
        } else if (num > second_largest) {
            second_largest = num;
        }
    }

    // Retorna el producto máximo de los dos números más grandes
    return largest * second_largest;
}

int main() {
    int n = 10;
    int min_value = 1; // Valor mínimo para los números aleatorios
    int max_value = 1000000; // Valor máximo para los números aleatorios
    int max_n = 1000000; // Valor máximo de n a probar

    // Empieza el cronómetro para medir el tiempo de ejecución
    auto start_time = std::chrono::high_resolution_clock::now();

    // Genera datos y calcula el producto máximo hasta que el tiempo supere 1 segundo
    while (n <= max_n) {
        generate_input_data(n, min_value, max_value); // Genera los datos de entrada para el valor actual de n

        // Mide el tiempo de ejecución para procesar el archivo de entrada
        auto start = std::chrono::high_resolution_clock::now();
        std::ifstream input_file("input.txt");
        int num;
        std::vector<int> input_data;
        input_file >> num;  // Lee el tamaño del vector (n)
        input_data.reserve(num);  // Reserva memoria para el número de elementos

        // Lee los números del archivo e ingrésalos en el vector
        while (input_file >> num) {
            input_data.push_back(num);
        }
        input_file.close();

        // Llama a la función para calcular el producto máximo de pares
        super_efficient_max_pairwise_product(input_data);

        // Detiene el cronómetro y calcula el tiempo transcurrido
        auto end = std::chrono::high_resolution_clock::now();
        double elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;

        // Muestra el tiempo que tomó para procesar el tamaño de n actual
        std::cout << "n = " << n << ", Time taken: " << elapsed_time << " seconds" << std::endl;

        // Si el tiempo transcurrido supera 1 segundo, detén el ciclo
        if (elapsed_time >= 1.0) {
            break;
        }

        // Duplicamos n para probar el siguiente tamaño
        n *= 2;
    }

    // Muestra el valor máximo de n que se puede procesar en 1 segundo
    std::cout << "Maximum n that can be solved in 1 second: " << n << std::endl;

    return 0;
}


Writing max_n_in_a_second_super_efficient.cpp


Test with inputs 1 and 2




In [174]:
!g++ -o max_n_in_a_second_super_efficient max_n_in_a_second_super_efficient.cpp
!time ./max_n_in_a_second_super_efficient < input1.txt
!time ./max_n_in_a_second_super_efficient < input2.txt

n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0 seconds
n = 10240, Time taken: 0.001 seconds
n = 20480, Time taken: 0.002 seconds
n = 40960, Time taken: 0.005 seconds
n = 81920, Time taken: 0.01 seconds
n = 163840, Time taken: 0.021 seconds
n = 327680, Time taken: 0.044 seconds
n = 655360, Time taken: 0.09 seconds
Maximum n that can be solved in 1 second: 1310720

real	0m0.373s
user	0m0.349s
sys	0m0.022s
n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0 seconds
n = 10240, Time taken: 0.001 s

Test and time it with input 3

In [175]:
!time ./max_n_in_a_second_super_efficient < input3.txt

n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0 seconds
n = 10240, Time taken: 0.001 seconds
n = 20480, Time taken: 0.002 seconds
n = 40960, Time taken: 0.005 seconds
n = 81920, Time taken: 0.01 seconds
n = 163840, Time taken: 0.021 seconds
n = 327680, Time taken: 0.043 seconds
n = 655360, Time taken: 0.09 seconds
Maximum n that can be solved in 1 second: 1310720

real	0m0.378s
user	0m0.351s
sys	0m0.024s


20 stress tests

In [176]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_n_in_a_second_super_efficient.cpp -o max_n_in_a_second_super_efficient

# Bucle para procesar los archivos desde input13.txt hasta input32.txt
for i in {13..32}
do
    # Ejecutar el programa con el archivo de entrada correcto
    ./max_n_in_a_second_super_efficient < input$i.txt > ctest$i.txt
    echo "Testing on input$i"

    # Medir el tiempo de ejecución
    time ./max_n_in_a_second_super_efficient < input$i.txt

    # Comparar la salida generada con la salida esperada
    echo ""

    # Verificar la diferencia entre el archivo de salida esperado y el generado
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [177]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Testing on input13
n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0 seconds
n = 10240, Time taken: 0.001 seconds
n = 20480, Time taken: 0.002 seconds
n = 40960, Time taken: 0.005 seconds
n = 81920, Time taken: 0.01 seconds
n = 163840, Time taken: 0.021 seconds
n = 327680, Time taken: 0.043 seconds
n = 655360, Time taken: 0.087 seconds
Maximum n that can be solved in 1 second: 1310720

real	0m0.371s
user	0m0.346s
sys	0m0.023s

n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0 seconds
n = 10240

## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [178]:
%%writefile super_efficient_vs_sorting.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>

// Función para generar un vector de entrada aleatorio
std::vector<int> generate_random_input(int n, int min_value, int max_value) {
    std::vector<int> input;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis(min_value, max_value);

    for (int i = 0; i < n; ++i) {
        input.push_back(dis(gen));
    }

    return input;
}

// Solución super eficiente
long long super_efficient_max_pairwise_product(const std::vector<int>& numbers) {
    long long largest = 0, second_largest = 0;

    // Encuentra los dos números más grandes
    for (int num : numbers) {
        if (num > largest) {
            second_largest = largest;
            largest = num;
        } else if (num > second_largest) {
            second_largest = num;
        }
    }

    return largest * second_largest;
}

// Solución basada en ordenamiento
long long sort_max_pairwise_product(const std::vector<int>& numbers) {
    std::vector<int> sorted_numbers = numbers;
    std::sort(sorted_numbers.begin(), sorted_numbers.end());
    int n = sorted_numbers.size();
    return (long long)sorted_numbers[n - 1] * sorted_numbers[n - 2];
}

int main() {
    const int num_tests = 100; // Número de pruebas
    const int n = 10;          // Tamaño del vector de entrada
    const int min_value = 1;   // Valor mínimo para los números aleatorios
    const int max_value = 1000; // Valor máximo para los números aleatorios

    bool error = false;

    // Realizar las pruebas
    for (int i = 0; i < num_tests; ++i) {
        // Genera una entrada aleatoria
        std::vector<int> input = generate_random_input(n, min_value, max_value);

        // Calcula el resultado de ambas soluciones
        long long super_efficient_result = super_efficient_max_pairwise_product(input);
        long long sort_result = sort_max_pairwise_product(input);

        // Compara los resultados
        if (super_efficient_result != sort_result) {
            std::cout << "Error: Outputs mismatch for test " << i + 1 << std::endl;
            std::cout << "Input: ";
            for (int num : input) {
                std::cout << num << " ";
            }
            std::cout << std::endl;
            std::cout << "Super-efficient result: " << super_efficient_result << std::endl;
            std::cout << "Sorting result: " << sort_result << std::endl;
            error = true;
            break;
        }
    }

    if (!error) {
        std::cout << "Pass: All tests passed successfully." << std::endl;
    }

    return 0;
}


Writing super_efficient_vs_sorting.cpp


In [179]:
#Create 100 random stress tests

!g++ stress_test_generator.cpp -o stress_test_generator
!./stress_test_generator

#Start index: 101 for file100
#numbers of tests: 100 (from 101 to 200)

Ingrese el índice inicial para los archivos de entrada: 101
Ingrese la cantidad de pruebas de estrés a crear: 100


In [180]:
%%writefile run_all_test.sh
#!/bin/bash
g++ super_efficient_vs_sorting.cpp -o super_efficient_vs_sorting

# Bucle para procesar los archivos desde input13.txt hasta input32.txt
for i in {101..200}
do
    # Ejecutar el programa con el archivo de entrada correcto
    ./super_efficient_vs_sorting < input$i.txt > ctest$i.txt
    echo "Testing on input$i"

    # Medir el tiempo de ejecución
    time ./super_efficient_vs_sorting < input$i.txt

    # Comparar la salida generada con la salida esperada
    echo ""

    # Verificar la diferencia entre el archivo de salida esperado y el generado
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [181]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Testing on input101
Pass: All tests passed successfully.

real	0m0.006s
user	0m0.005s
sys	0m0.001s

Pass: All tests passed successfully.
Testing on input102
Pass: All tests passed successfully.

real	0m0.006s
user	0m0.005s
sys	0m0.001s

Pass: All tests passed successfully.
Testing on input103
Pass: All tests passed successfully.

real	0m0.006s
user	0m0.005s
sys	0m0.001s

Pass: All tests passed successfully.
Testing on input104
Pass: All tests passed successfully.

real	0m0.007s
user	0m0.004s
sys	0m0.002s

Pass: All tests passed successfully.
Testing on input105
Pass: All tests passed successfully.

real	0m0.006s
user	0m0.004s
sys	0m0.002s

Pass: All tests passed successfully.
Testing on input106
Pass: All tests passed successfully.

real	0m0.009s
user	0m0.006s
sys	0m0.000s

Pass: All tests passed successfully.
Testing on input107
Pass: All tests passed successfully.

real	0m0.008s
user	0m0.005s
sys	0m0.001s

Pass: All tests passed successfully.
Testing on input108
Pass: All tests passe